PublicMapping/districtbuilder-classic
# Feature/js/setup dev environment

Overview
-----

This PR takes steps toward having a working local instance of the application.

Its main actions are:

- run the django server using gunicorn
- remove unused dependencies
- move files around as needed
- move the django server into a container
- move the database and redis into containers
- upgrade a bunch of dependencies (django, redis, django-redis, postgres, postgis...)
- use django tools that may not have existed in django 1.4 (like `django.core.cache` and the `{% static ... %}` interpolator)
- upgrade django settings (include a `CACHES` section for the cache, e.g.)
- modify url routing
- kill a _ton_ of trailing whitespace (but by no means all of it -- it's dead everywhere I visited I think but I probably forgot to do it in some places)

It's _huge_, and therefore pretty much impossible to test completely.

In the application's current state, you can hit the landing page and all of the links appear to work, but trying to do (basically) anything fails. Pretty much the only thing that works is that you can try to log in as a guest, and then the app will successfully tell you that you've failed to login (:fire:). This prevents exercising some of the application's features.

Notes
-----

The "spread firefox support" button seems not to exist anywhere in mozilla anymore. I didn't remove it because I wasn't sure if we wanted to go hunting for a replacement.

There are a few things that I either haven't exercised or haven't fixed:

- geoserver remains untested. I'll integrate the container into this work in when I get a PR up for it
- `gettext` isn't working at all. That seems really bad! (#459)
- Something is generically wrong with basically all of the views, where they complain about `No reverse lookup found` or believe they have syntax errors in the templates or who knows what else. Digging through template inheritance seemed like a possible black hole (#460)

Testing
-----

- `docker-compose up`
- Test the cache connection:
  - `docker-compose exec django bash`
  - `./manage.py shell`
  - run this:
```python
from django.core.cache import caches
cache = caches['default']
cache.get('foo')
```
    
- you should get nothing instead of an error
- Test the database connection
  - from inside the django container, `./manage.py dbshell`
  - `\d`
  - you should see a bunch of tables
  - `\d redistricting_district`
  - you should see something that looks like a reasonable data model
- test the frontend
  - navigate to `localhost:8080`
  - the only error you should see in the console should be that the frontend doesn't know what `gettext` is

Closes (or at least convincingly starts on) #455 
Closes #458
